Support the telegraf receiver plugin module

soander opened the pull request:
<!--
    ⚠️ Please make sure to read this template first, pull requests that don't accord with this template
    maybe closed without notice.
    Texts surrounded by `<` and `>` are meant to be replaced by you, e.g. <framework name>, <issue number>.
    Put an `x` in the `[ ]` to mark the item as CHECKED. `[x]`
-->

<!-- ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👇 ====
### Fix <bug description or the bug issue number or bug issue link>
- [ ] Add a unit test to verify that the fix works.
- [ ] Explain briefly why the bug exists and how to fix it.
     ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👆 ==== -->

<!-- ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👇 ====
### Improve the performance of <class or module or ...>
- [ ] Add a benchmark for the improvement, refer to [the existing ones](https://github.com/apache/skywalking/blob/master/apm-commons/apm-datacarrier/src/test/java/org/apache/skywalking/apm/commons/datacarrier/LinkedArrayBenchmark.java)
- [ ] The benchmark result.
```text
<Paste the benchmark results here>
```
- [x] Links/URLs to the theory proof or discussion articles/blogs. <links/URLs here>
     ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👆 ==== -->


### <Support the telegraf receiver plugin module>
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [X] Update the documentation to include this new feature.
- [X] Tests(including UT, IT, E2E) are added to verify the new feature.
- [X] If it's UI related, attach the screenshots below.



- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [X] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).


**Description**
- 
It's a new way to monitor different servers' metrics. This module is telegraf receiver plugin. It mainly used to receive, process and send telegraf's collected metrics, such as cpu, memory, disk etc.


**Design**
-
- Modularized the telegraf receiver plugin by Java SPI
- Defined the telegraf receiver plugin module name, provider, handler
- Made the configuration file to set different MAL expressions

**Step**
-
- Collected servers’ metrics such as system load, cpu usage, memory usage by InfluxDB telegraf
- Received collected `JSON` format metrics in HTTP messages by RESTful API
- Converted JSON format metrics to Skywalking's metrics format by Jackson API
- Sent converted metrics to meter system, did the aggregation and stored it to database
- Queried stored metrics by GraphQL and displayed on Skywalking's UI

**Attention**
-
- The module uses `HTTP` to receive telegraf's metrics, so users should set `[[outputs.http]]` in telegraf.conf file
- The module **only** processes telegraf's `JSON` format metrics, so users must configure `data_format = "json"` in telegraf.conf file
- The default `json_timestamp_units` is `second`, the module **only** process `second` timestamp unit. If you configure `json_timestamp_units`, `json_timestamp_units = "1s"` is feasible.

**Example**
-
The following screenshot shows the detail query information.
<img width="1090" alt="image" src="https://user-images.githubusercontent.com/79000735/190608100-7d9aaf1c-df59-47c4-ba39-7b8ecd31bce0.png">

wu-sheng commented on oap-server/server-starter/src/main/resources/telegraf-rules/default.yaml:
```diff
@@ -0,0 +1,50 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+expSuffix: service(['host'], Layer.OS_LINUX)
```
Is this file working well as same as OTEL Linux monitoring? Could we activate them at the same time?

soander replied:
> Is this file working well as same as OTEL Linux monitoring? Could we activate them at the same time?

I am not sure whether they can activate at the same time, I will test it.

wu-sheng replied:
If it can't for now, please let me know. Unless there are differences at aggregation function level, or downsampling. We should be able to.

Otherwise, this configurations(rules) should be in default release. Because there is no dashboard for it.

wu-sheng reviewed (changes requested):
Several important things should be added.
1. Explain how telegraf metric configuration working with OTEL OTLP/OC receiver
2. `application.yml` should be changed, otherwise, this module is never able to be activated
3. Documents about telegraf should be added.
4. e2e for this new format should be added. At least for your added rules.

wankai123 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/handler/TelegrafServiceHandler.java:
```diff
@@ -0,0 +1,128 @@
+
+        fields.forEach((key, value) -> {
+            Sample.SampleBuilder builder = Sample.builder();
+            if (value instanceof String) return;
```
Hi, could you provide some JSON data samples? I want to know more details about the convert logic. Such as why the value might be String.

soander replied:
> Hi, could you provide some JSON data samples? I want to know more details about the convert logic. Such as why the value might be String.

The following data is JSON data samples:
`{"metrics":[{"fields":{"available":6047739904,"available_percent":35.41215070500567,"total":17078149120,"used":11030409216,"used_percent":64.58784929499433},"name":"mem","tags":{"host":"DESKTOP-S9HLAV6"},"timestamp":1663391390},{"fields":{"usage_guest":0,"usage_guest_nice":0,"usage_idle":95.32710280373831,"usage_iowait":0,"usage_irq":0.3115264797507788,"usage_nice":0,"usage_softirq":0,"usage_steal":0,"usage_system":1.7133956386292835,"usage_user":2.64797507788162},"name":"cpu","tags":{"cpu":"cpu0","host":"DESKTOP-S9HLAV6"},"timestamp":1663391390}]}`

The convert logic is build a sample by iterating every pair of key and value. Then adding the single simple to the sample list. Maybe the value will never be String, so the line of code is probably useless.

wankai123 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/handler/TelegrafServiceHandler.java:
```diff
@@ -0,0 +1,128 @@
+            Sample.SampleBuilder builder = Sample.builder();
+            if (value instanceof String) return;
+            Sample sample = builder.name(name + "_" + key)
+                    .timestamp(timestamp * 1000L)
```
can you confirm the timestamp unit is `second` by default?
I found a doc: https://docs.influxdata.com/telegraf/v1.23/data_formats/output/json/
maybe the timestamp unit is configurable.

soander replied:
> can you confirm the timestamp unit is `second` by default? I found a doc: https://docs.influxdata.com/telegraf/v1.23/data_formats/output/json/ maybe the timestamp unit is configurable.

Yeah, I can confirm that the default timestamp unit is `second`. Because I just set the `data_format = "json"`, and didn't configure the `json_timestamp_units`. According to output timestamp result, the default timestamp is `second`.
The following screenshot is my configuration.
<img width="538" alt="image" src="https://user-images.githubusercontent.com/79000735/190842615-9c08fc66-70be-405f-892c-8958de710c2b.png">

So what should I do to handle different timestamp unit, such as `"1ns", "1us", "1ms", "10ms"`? I will appreciate your guidance!

wu-sheng replied:
We should have a configuration to match telegraf configuration about time precision.

In the MAL, the timestamp(from Java spec) is ms, please refer to the Java spec

kezhenxu94 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/TelegrafReceiverProvider.java:
```diff
@@ -0,0 +1,93 @@
+
+    @Override
+    public ModuleConfig createConfigBeanIfAbsent() {
+        moduleConfig = new TelegrafModuleConfig();
+        return moduleConfig;
```
This is not `create if absent` (`createConfigBeanIfAbsent`)

kezhenxu94 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/config/TelegrafConfigs.java:
```diff
@@ -0,0 +1,71 @@
+                    try (Reader r = new FileReader(f)) {
+                        return new Yaml().loadAs(r, TelegrafConfig.class);
+                    } catch (IOException e) {
+                        log.warn("Reading file {} failed", f, e);
```
Please throw module start exception, this is a fatal error

soander replied:
> Please throw module start exception, this is a fatal error

I set the `loadConfigs` method throws the `ModuleStartException` , should I change it?
![image](https://user-images.githubusercontent.com/79000735/190843723-2cbbfe34-9552-424c-a370-28d31f35f2c4.png)

kezhenxu94 replied:
> > Please throw module start exception, this is a fatal error
> 
> I set the `loadConfigs` method throws the `ModuleStartException` , should I change it? ![image](https://user-images.githubusercontent.com/79000735/190843723-2cbbfe34-9552-424c-a370-28d31f35f2c4.png)

I mean you should throw a `ModuleStartException` here, logging it is not enough.

```suggestion
                        throw new ModuleStartException(e);
```

kezhenxu94 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/handler/TelegrafServiceHandler.java:
```diff
@@ -0,0 +1,128 @@
+
+    @Post("/telegraf")
+    public Commands collectData(String jsonInfo) {
+        TelegrafData telegrafData;
```
```suggestion
```

kezhenxu94 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/handler/TelegrafServiceHandler.java:
```diff
@@ -0,0 +1,128 @@
+            // Read each metrics json and convert it to Sample
+            JsonNode metrics = node.get("metrics");
+            for (JsonNode m : metrics) {
+                telegrafData = mapper.convertValue(m, TelegrafData.class);
```
```suggestion
                TelegrafData telegrafData = mapper.convertValue(m, TelegrafData.class);
```

kezhenxu94 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/handler/TelegrafServiceHandler.java:
```diff
@@ -0,0 +1,128 @@
+        TelegrafData telegrafData;
+        try (HistogramMetrics.Timer ignored = histogram.createTimer()) {
+            List<Sample> allSamples = new ArrayList<>();
+            ObjectMapper mapper = new ObjectMapper();
```
Extract this to a field, the mapper should not be instantiated every time.

kezhenxu94 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/handler/pojo/TelegrafData.java:
```diff
@@ -0,0 +1,42 @@
+@AllArgsConstructor
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class TelegrafData {
+    @JsonProperty("fields")
+    private Map<String, Object> fields;
+    @JsonProperty("name")
+    private String name;
+    @JsonProperty("tags")
+    private Map<String, String> tags;
+    @JsonProperty("timestamp")
+    private long timestamp;
```
If the json field names are the same as the Java field names, `@JsonProperty` is redundant

kezhenxu94 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/handler/TelegrafServiceHandler.java:
```diff
@@ -0,0 +1,128 @@
+    **/
+    public List<Sample> convertTelegraf(TelegrafData telegrafData) {
+
+        List<Sample> sampleList = new ArrayList<>(Collections.emptyList());
```
```suggestion
        List<Sample> sampleList = new ArrayList<>();
```

kezhenxu94 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/handler/TelegrafServiceHandler.java:
```diff
@@ -0,0 +1,128 @@
+        try (HistogramMetrics.Timer ignored = histogram.createTimer()) {
+            List<Sample> allSamples = new ArrayList<>();
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode node = mapper.readTree(jsonInfo);
```
Can you rename the original `TelegrafData` class to `TelegrafDatum`, and define a class `TelegrafData` to contain a field `List<TelegrafDatum> metrics`, so you don't need to `readTree` here, and just declare the method `public Commands collectData(TelegrafData telegrafData) {`

soander replied:
> Can you rename the original `TelegrafData` class to `TelegrafDatum`, and define a class `TelegrafData` to contain a field `List<TelegrafDatum> metrics`, so you don't need to `readTree` here, and just declare the method `public Commands collectData(TelegrafData telegrafData) {`

I change code according to your suggestion. Everything works fine.

This is `TelegrafData` class.
![image](https://user-images.githubusercontent.com/79000735/191666192-76d80f30-eed9-4569-ae26-3619578f0b98.png)

This is `public Commands collectData(TelegrafData telegrafData)` method.
![image](https://user-images.githubusercontent.com/79000735/191666960-bf395584-b0db-494f-b02e-3ed922c7863c.png)

Is this modification good? If there are still problems, what else do I need to fix? Thank you for your guidance!

kezhenxu94 replied:
Not sure whether it's necessary that class `TelegrafData` should implement `RequestConverterFunction`, this is a standard POJO and the Armeria should be able to deserialize it correctly

soander replied:
> Not sure whether it's necessary that class `TelegrafData` should implement `RequestConverterFunction`, this is a standard POJO and the Armeria should be able to deserialize it correctly

I set the class `TelegrafData` like this:
![image](https://user-images.githubusercontent.com/79000735/192081885-8811cae0-2432-4184-b197-c074d451c68f.png)

The method shows this **error**.
![image](https://user-images.githubusercontent.com/79000735/192082071-73e300a9-b1b5-4090-a395-a69ee0f17b47.png)

After I removed the code `@RequestConverter(TelegrafData.class)`, the oap-server can be run, but the class `TelegrafServiceHandler` can't work, and pops up **error** info:
![image](https://user-images.githubusercontent.com/79000735/192081965-2df5f991-e30c-4202-a709-c21b17785985.png)

According to the Armeria document, the POJO class should implement `RequestConverterFunction` that can be used in the `@RequestConverter` and `@RequestObject` annotations. You can check the section **Conversion between an HTTP message and a Java object** in this document [https://armeria.dev/docs/server-annotated-service/](url) or the part **Write a request converter** in this tutorials [https://armeria.dev/tutorials/rest/blog/implement-create/](url).

Maybe the class `TelegrafData` can be optimized, but I think class `TelegrafData `must implement `RequestConverterFunction` for converting JSON object to JAVA object.

wu-sheng commented:
Any update as has been reviewed for a week.

soander commented:
> @soander Could you put a summary about which are changed from last comment?

@wu-sheng Compared with the last comment:

**Code file**
-
- Changed code according to **suggested changes** like `createConfigBeanIfAbsent` method in the `TelegrafReceiverProvider` class, `ModuleStartException` in the `TelegrafConfigs` class
- Refactor code about **converting** JSON format telegraf data to Java POJO, created new class `TelegrafDatum` and rewrite class 'TelegrafData'
- Removed redundant `@JsonProperty` annotations in the class `TelegrafDatum`

**Configuration file**
-
- Add the telegraf  module configuration in the `application.yml` to activate it

**Document file**
- 
- Update the `changes.md`
- Added the `Design`, `Step`, `Attention` in PR description

wu-sheng commented on docs/en/changes/changes.md:
```diff
@@ -20,6 +20,7 @@
   in `no-init` mode.
 * Make sure instance list ordered in TTL processor to avoid TTL timer never runs.
 * Support monitoring PostgreSQL slow SQLs.
+* Support Telegraf receiver plugin module.
```
```suggestion
* Support Telegraf receiver plugin module.
* Support monitoring Linux OS through Telegraf.
```

Also, you should update two docs, and update change logs accordingly too.

wu-sheng reviewed (commented):
1. Please rename [server-starter/src/main/resources/telegraf-rules/default.yaml](https://github.com/apache/skywalking/pull/9620/files#diff-154f3f77f17bc71732fa1168c0771d9813b345587798e89cc7e1fa3365053f24) to a meaningful name, due to this is actually analyzing Linux OS metrics?
2. You should refer [OTEL receiver doc](https://skywalking.apache.org/docs/main/next/en/setup/backend/opentelemetry-receiver/), and add a new Telegraf doc.
3. You should update [Linux Monitoring Dashboard doc](https://skywalking.apache.org/docs/main/next/en/setup/backend/backend-vm-monitoring/). <1> Describe there are two ways in default could monitor Linux OS, one is existing OTEL, and the other is this new.  <2> Re-organize the doc to separate these two ways.
4. Code style check fails, please make sure all CI could passed.
5. e2e is still missing. Please ref the existing one
https://github.com/apache/skywalking/blob/bf4cf9d8a026b82693066719c1607eb79d35a844/.github/workflows/skywalking.yaml#L549-L550

wu-sheng commented:
One important thing, please make sure your codes could pass all code checks, UT, and e2e tests before request reviewing. Otherwise, we may even read the wrong codes, which doesn't make sure to make many reviewers put time into a PR.

soander commented:
> One important thing, please make sure your codes could pass all code checks, UT, and e2e tests before request reviewing. Otherwise, we may even read the wrong codes, which doesn't make sure to make many reviewers put time into a PR.

I'm sorry for the inconvenience. How can I do the code checks, UT, and e2e tests locally? Thank you for your guidance！

wu-sheng commented:
It is better you could ask for your mentor, as this is a Summer 2022 Program.

Generally, there are 
- All tests are controlled through GitHub Action, https://github.com/apache/skywalking/tree/master/.github/workflows
- About e2e tests, https://skywalking.apache.org/docs/main/next/en/guides/readme/#end-to-end-tests-e2e

For merging this in upstream, testing is a very important, also time costing task. It may take more time than you wrote this feature. But this is how the open-source community works, in order to collaborate with people remotely.
Be patient.

soander commented:
> 

I have contacted my mentor, and I will check documents. Maybe I made some inconvenience because I am a newbie.
Anyway, thank you again and sorry for some mistakes.

wu-sheng commented:
I recommend you `asking for your mentor`, because I think he may have more time than me to help you out with the env and setup issues.
This is not an issue from community side, we just have to wait for your tests ready.

wu-sheng commented:
Marked this as `no update`, as we can't have any feedback for two weeks. I am not sure whether this is still on going or still prepares for upstream review.

liuhaoyang commented:
> Marked this as `no update`, as we can't have any feedback for two weeks. I am not sure whether this is still on going or still prepares for upstream review.
This PR are still being revised. I will help him to add the e2e tests ..

wu-sheng commented:
4 files left for this PR? What happens?

soander commented:
> 4 files left for this PR? What happens?

I made a wrong push...

wu-sheng commented:
What is the current status of this PR?

wu-sheng commented on docs/en/setup/backend/backend-vm-monitoring.md:
```diff
@@ -3,38 +3,63 @@ SkyWalking leverages Prometheus node-exporter to collect metrics data from the V
+3. Config SkyWalking [Telegraf receiver](telegraf-receiver.md).
+
 ## Supported Metrics
+**For OpenTelemetry receiver:**
```
What does this mean? OpenTelemetry receiver and Telegraf receiver don't share the metrics?

wu-sheng replied:
If they are same, keep dataflow doc is fine, but you don't have to list two sheets for the same metrics. Right?

soander replied:
> If they are same, keep dataflow doc is fine, but you don't have to list two sheets for the same metrics. Right?

But the metrics name and data source are different. If the telegraf receiver sheet is redundant I can delete it.

wu-sheng replied:
Metrics name for UI, or metrics name from telegraf?

soander replied:
> Metrics name for UI, or metrics name from telegraf?

Both metrics names are different. For example, telegraf metrics' `mem_total` shows `mem_total_info` in UI, but OpenTelemetry metrics' `meter_vm_memory_total` shows `memory_total` in UI.

wu-sheng commented:
Besides e2e, you should add UTs(Unit Tests) for your format transfer, it could cover more cases easily and help the reviewers to understand the codes.

liuhaoyang commented on test/e2e-v2/cases/vm/telegraf/telegraf.conf:
```diff
@@ -0,0 +1,9559 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
`telegraf.conf` can be made shorter. You should delete unused configs instead of just commenting them out.

soander replied:
> `telegraf.conf` can be made shorter. You should delete unused configs instead of just commenting them out.
Ok, get it. I will delete redundant configs.

liuhaoyang commented:
@wu-sheng I see all the e2e tests return failure, what could be causing this ?

wu-sheng commented:
https://github.com/apache/skywalking/actions/runs/3288941909

You could find logs from this page, at the bottom of the page. Most likely, the package of this branch can't boot up.

wu-sheng commented:
Has your mentor reviewed, checked and tested this?

liuhaoyang commented:
> Has your mentor reviewed, checked and tested this?

Yes. I have reviewed and performed the tests.

wu-sheng commented on .mvn/wrapper/maven-wrapper.properties:
```diff
@@ -5,14 +5,14 @@
-# 
+#
 #   http://www.apache.org/licenses/LICENSE-2.0
-# 
+#
 # Unless required by applicable law or agreed to in writing,
 # software distributed under the License is distributed on an
 # "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
 distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.8.4/apache-maven-3.8.4-bin.zip
-wrapperUrl=https://repo.maven.apache.org/maven2/org/apache/maven/wrapper/maven-wrapper/3.1.0/maven-wrapper-3.1.0.jar
+wrapperUrl=https://repo.maven.apache.org/maven2/org/apache/maven/wrapper/maven-wrapper/3.1.0/maven-wrapper-3.1.0.jar
```
There are unexpected changes.

wu-sheng commented:
I still noticed there are several mentioned things not been fixed.

1. There is no unit test to cover telegraf format converting. `TelegrafServiceHandler` is a HTTP handler. What it received is an HTTP text, what it looks like, and how it could be converted to SampleFamily should be well tested, including legal text or illegal text cases.
2. You don't example what is the relationship between this VM monitoring and OpenTelemetry's Linux monitoring. 
3. In [backend-vm-monitoring.md](https://github.com/apache/skywalking/pull/9620/files#diff-9b751b71ba658be11f59887b6462a97bd3675a181c3cd3fcf7d5c4e3d452c295), you mentioned, there are two ways to support VM monitoring. But they seem not the same, such as not using the same metric name. So, how the dashboard would work? In the document, there is not any document mentioning that.

I still can't follow the whole picture of this feature implementation. Could any of you explain it? We can't deliver this to the community with this confusion. Especially, `telegraf-rules/vm.yaml` is included as a part of the official release.

All default activated MAL scripts match with the dashboard settings, as well as the document description about the data flow.

soander commented:
> I still noticed there are several mentioned things not been fixed.
> 
> 1. There is no unit test to cover telegraf format converting. `TelegrafServiceHandler` is a HTTP handler. What it received is an HTTP text, what it looks like, and how it could be converted to SampleFamily should be well tested, including legal text or illegal text cases.
> 2. You don't explain what is the relationship between this VM monitoring and OpenTelemetry's Linux monitoring.
> 3. In [backend-vm-monitoring.md](https://github.com/apache/skywalking/pull/9620/files#diff-9b751b71ba658be11f59887b6462a97bd3675a181c3cd3fcf7d5c4e3d452c295), you mentioned, there are two ways to support VM monitoring. But they seem not the same, such as not using the same metric name. So, how the dashboard would work? In the document, there is not any document mentioning that.
> 
> I still can't follow the whole picture of this feature implementation. Could any of you explain it? We can't deliver this to the community with this confusion. Especially, `telegraf-rules/vm.yaml` is included as a part of the official release.
> 
> All default activated MAL scripts match with the dashboard settings, as well as the document description about the data flow.

 1. I will write the HTTP text converting to SampleFamily unit test.
 2. Telegraf and OpenTelemetry are two different metrics collection methods, both of them can collect linux system metrics and display it on the UI.
 3.  The metrics are collected by telegraf can also be displayed on the VM monitoring dashboard. The default telegraf MAL will convert the telegraf metrics into VM metrics.

liuhaoyang commented:
> I still noticed there are several mentioned things not been fixed.
> 
> 1. There is no unit test to cover telegraf format converting. `TelegrafServiceHandler` is a HTTP handler. What it received is an HTTP text, what it looks like, and how it could be converted to SampleFamily should be well tested, including legal text or illegal text cases.
> 2. You don't explain what is the relationship between this VM monitoring and OpenTelemetry's Linux monitoring.
> 3. In [backend-vm-monitoring.md](https://github.com/apache/skywalking/pull/9620/files#diff-9b751b71ba658be11f59887b6462a97bd3675a181c3cd3fcf7d5c4e3d452c295), you mentioned, there are two ways to support VM monitoring. But they seem not the same, such as not using the same metric name. So, how the dashboard would work? In the document, there is not any document mentioning that.
> 
> I still can't follow the whole picture of this feature implementation. Could any of you explain it? We can't deliver this to the community with this confusion. Especially, `telegraf-rules/vm.yaml` is included as a part of the official release.
> 
> All default activated MAL scripts match with the dashboard settings, as well as the document description about the data flow.

This PR contains two features
1. Support telegraf-receiver to receive data in telegraf json format and write it to storage
2. Provide the default MAL configuration, and parse the CPU, memory and system metrics reported by telegraf into vm meter, so that these metrics can be displayed on the linux monitoring page. (No disk and network metrics are provided for the time being
For 1, users can use their own MAL to parse more telegraf metrics.

kezhenxu94 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/config/TelegrafConfigs.java:
```diff
@@ -0,0 +1,75 @@
+                        return new Yaml().loadAs(r, TelegrafConfig.class);
+                    } catch (IOException e) {
+                        log.warn("Reading file {} failed", f, e);
+                        try {
+                            throw new ModuleStartException(e.getMessage(), e);
+                        } catch (ModuleStartException ex) {
+                            return null;
+                        }
```
What the meaning here 🤯

soander replied:
It's an error, I already fix it.

soander replied:
> What the meaning here 🤯

It's an error, I already fix it.

wu-sheng commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/test/java/org/apache/skywalking/oap/server/receiver/telegraf/JSONConvertTest.java:
```diff
@@ -0,0 +1,126 @@
+
+import static org.junit.Assert.assertEquals;
+
+public class JSONConvertTest {
```
Could you check `ZabbixMetricsTest` and `ZabbixProtocolHandlerTest` about how to do UT? What we need to test is that, we have a JSON text(same format as telegraf reported, that could link to a doc), then, the UT calls some methods in your handler(or something here). Then we could verify the sample family with expectations.

In the current one, you seem using a mock class to test another mock class.

cc @kezhenxu94

wu-sheng commented on oap-server/server-starter/src/main/resources/ui-initialized-templates/os_linux/linux-service.json:
```diff
@@ -156,10 +156,12 @@
           "metrics": [
-            "meter_vm_filesystem_percentage"
+            "meter_vm_filesystem_percentage",
+            "meter_vm_filesystem_percentage_telegraf"
           ],
           "metricTypes": [
-            "readLabeledMetricsValues"
+            "readLabeledMetricsValues",
+            "readMetricsValues"
```
@Fine0830 I noticed this kind of hybrid setting, one for labeled metrics, the other for a single value. Is this supported in the plan?

@soander Please post screenshot about how it looks like in OpenTelemetry and Telegraf cases.

liuhaoyang replied:
I don't think these two metrics will appear at the same time in the same dashboard. The two indicators are only for compatibility with the data that the chart can query.

wu-sheng commented on apm-dist/src/main/assembly/binary.xml:
```diff
@@ -69,6 +69,7 @@
                 <include>ui-initialized-templates/*/*.json</include>
                 <include>lal/*</include>
                 <include>log-mal-rules/*</include>
+                <include>telegraf-rules/*.yaml</include>
```
```suggestion
                <include>telegraf-rules/*</include>

```
Why do we need a `yaml` suffix? I assume there is no other files in the folder, right?

wu-sheng commented on oap-server/server-starter/src/main/resources/telegraf-rules/vm.yaml:
```diff
@@ -0,0 +1,56 @@
+    exp: swap_free
+  - name: memory_swap_total
+    exp: cpu_swap_total
+  - name: memory_swap_percentage
+    exp: 100 - ((swap_free * 100) / cpu_swap_total)
```
Percentage usually uses 0.01% precision. And UI would set `/ 100` to format the two decimal places.

Could you check what is your case?

wu-sheng commented on docs/en/setup/backend/telegraf-receiver.md:
```diff
@@ -0,0 +1,45 @@
+# Telegraf receiver
```
Metrics receiver doc should be hosted in the menu like this

<img width="350" alt="image" src="https://user-images.githubusercontent.com/5441976/198949306-a7e2e327-4458-4d15-88ac-36fbe0b1ce14.png">

You should update the menu.yml in the docs folder.

wu-sheng commented:
The document seems should be polished after the code's part is ready.

kezhenxu94 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/config/TelegrafConfigs.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.stream.Collectors;
+
+@Slf4j
+public class TelegrafConfigs {
```
This file is duplicate of `org.apache.skywalking.oap.meter.analyzer.prometheus.rule.Rules`

kezhenxu94 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/config/TelegrafConfig.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.List;
+
+@Data
+public class TelegrafConfig implements MetricRuleConfig {
```
This class is duplicate of `org.apache.skywalking.oap.meter.analyzer.prometheus.rule.Rule`

kezhenxu94 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/config/TelegrafConfig.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Data
+    @NoArgsConstructor
+    public static class Rule implements RuleConfig {
```
This class is duplicate of `org.apache.skywalking.oap.meter.analyzer.prometheus.rule.MetricsRule`

kezhenxu94 commented on oap-server/server-starter/src/main/resources/telegraf-rules/vm.yaml:
```diff
@@ -0,0 +1,56 @@
+  - name: cpu_total_percentage
+    exp: cpu_usage_user.tagEqual('cpu' , 'cpu-total')
+  - name: cpu_average_used_telegraf
+    exp: cpu_usage_user.tagEqual('cpu' , 'cpu-total')
```
The `exp` is exactly the same as `cpu_total_percentage`, may be it's something like `cpu_usage_user.tagEqual('cpu' , 'cpu-used')`? Can you recheck

kezhenxu94 reviewed (commented):
Just go through the PR and submit the comments for fixing, will take another look soon.

mrproliu commented on oap-server/server-starter/src/main/resources/telegraf-rules/vm.yaml:
```diff
@@ -0,0 +1,56 @@
+
+    #node filesystem
+  - name: filesystem_percentage_telegraf
+    exp: disk_used_percent
```
I have done a local test with the telegref, I think we need to use some aggregation function to aggregate the metrics: 
1. The telegref will count the metrics every second and batch upload them. 
![image](https://user-images.githubusercontent.com/3417650/198979802-66e7811b-f2ca-43ef-bdba-fbc4ee2d1f1c.png)
2. The telegref report data with multiple labels. for example, the dist used percent meter will contain metrics with multiple devices. 
![image](https://user-images.githubusercontent.com/3417650/198981390-ed44b9f5-61eb-4f02-84ab-992c368b2ee1.png)

liuhaoyang replied:
for 1, the [interval](https://github.com/influxdata/telegraf/blob/master/docs/CONFIGURATION.md#intervals) is configurable in telegraf. In the telegraf.conf of the e2e test, in order to obtain data faster, configure the interval to `1s`

wu-sheng replied:
Too much data could cause tests unstable because GHA only provides 3U8G to the whole test.

For the screenshot, the biggest issue is, SampleFamily should have a concept of the group for the bulk. We should not face one SampleFamily have metrics belonging to the same metric and labels, but with multiple timestamps.

Otherwise, the aggregation result would be strange and unpredictable.

Could you share how you process this?

liuhaoyang replied:
It is feasible to change the collection interval of telegraf to `10s` or `30s`

liuhaoyang replied:
The reason for this problem is the wrong config of telegraf, the collection interval is much shorter than the flusher interval.

wu-sheng commented on docs/menu.yml:
```diff
@@ -115,6 +115,8 @@ catalog:
             path: "/en/guides/backend-oal-scripts"
           - name: "OpenTelemetry Metrics"
             path: "/en/setup/backend/opentelemetry-receiver"
+          - name: "Telegraf Metrics"
+            path: "/en/setup/backend/telegraf-receiver"
```
Move this next to the Meter Analysis.

liuhaoyang commented:
There seems to be a problem with the e2e test case of vm-telegraf, @soander  can you fix it?
Alse, I used the latest commit to test, and the dashboards can be displayed normally (I did not find the Filefd Allocated indicator collection plugin in telegraf, so this chart has no data display for the time being
![image](https://user-images.githubusercontent.com/10701915/199729883-37028eb3-b289-458b-b76b-2f5cce2de8e7.png)

Waiting for your review @wu-sheng @mrproliu @kezhenxu94 , thanks again

mrproliu reviewed (dismissed):
LGTM

kezhenxu94 commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/TelegrafReceiverProvider.java:
```diff
@@ -0,0 +1,110 @@
+    private List<Rule> configs;
+    private TelegrafModuleConfig moduleConfig;
+
+    public TelegrafReceiverProvider() {
+        this.moduleConfig = new TelegrafModuleConfig();
+    }
+
```
No need

```suggestion
```

kezhenxu94 reviewed (dismissed):
LGTM overall

wu-sheng commented:
The one thing that needs to be fixed is this, 

```
For the screenshot, the biggest issue is, SampleFamily should have a concept of the group for the bulk. We should not face one SampleFamily have metrics belonging to the same metric and labels, but with multiple timestamps.

Otherwise, the aggregation result would be strange and unpredictable.
```

wu-sheng commented on oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/handler/TelegrafServiceHandler.java:
```diff
@@ -0,0 +1,129 @@
+            allSamples.addAll(samples);
+        }
+
+        // Grouping all samples by their name, then build sampleFamily
```
This policy seems not right. The samples should be grouped by bulk/sampling ID. 
Because 
1. Metrics in the same name with different timestamps should belong to different sample families.
2. Different metrics(name) collected in the same timestamp(one collecting) should belong to the same sample family. You could check the MAL doc, which mentioned metrics should be calculated through one formula, ref to https://skywalking.apache.org/docs/main/next/en/concepts-and-designs/mal/#binary-operators, `instance_trace_analysis_error_count / instance_trace_count`

soander replied:
> This policy seems not right. The samples should be grouped by bulk/sampling ID. Because
> 
> 1. Metrics in the same name with different timestamps should belong to different sample families.
> 2. Different metrics(name) collected in the same timestamp(one collecting) should belong to the same sample family. You could check the MAL doc, which mentioned metrics should be calculated through one formula, ref to https://skywalking.apache.org/docs/main/next/en/concepts-and-designs/mal/#binary-operators, `instance_trace_analysis_error_count / instance_trace_count`

Got it. I will modify this part logic.

soander replied:
> This policy seems not right. The samples should be grouped by bulk/sampling ID. Because
> 
> 1. Metrics in the same name with different timestamps should belong to different sample families.
> 2. Different metrics(name) collected in the same timestamp(one collecting) should belong to the same sample family. You could check the MAL doc, which mentioned metrics should be calculated through one formula, ref to https://skywalking.apache.org/docs/main/next/en/concepts-and-designs/mal/#binary-operators, `instance_trace_analysis_error_count / instance_trace_count`

I already fix the issue by grouping sample family by name and timestamp, and extra UTs have been tested.

wu-sheng reviewed (dismissed):
LGTM once tests are passed.

wankai123 commented on oap-server/server-starter/src/main/resources/telegraf-rules/vm.yaml:
```diff
@@ -0,0 +1,72 @@
+  - name: cpu_total_percentage
+    exp: cpu_usage_active.tagEqual('cpu', 'cpu-total')
+  - name: cpu_average_used
+    exp: cpu_usage_active.tagNotEqual('cpu', 'cpu-total').avg(['host', 'cpu'])
```
Maybe this expression does not calculate the `The percentage usage of the CPU core in each mode`  as the doc explains. 
If this metric has a different meaning by using the telegraf, I think you should better notice it in the doc.

wu-sheng replied:
@soander Please take a look. We may need some polish.

wankai123 commented:
The others LGTM

wu-sheng commented:
One more thing, `telegraf` should be added as a label for our repo, https://github.com/apache/skywalking/blob/master/.asf.yaml

wu-sheng commented:
@liuhaoyang Please make sure squash merge log is short and clear.
You can see there are a lot of noices in the last commit.
